### PR TITLE
Make Ambassador Pro installable from helm chart

### PR DESCRIPTION
**Pull request by iNoahNothing**
#### What this PR does / why we need it:

Adds option to install Ambassador Pro from the Helm chart

#### Which issue this PR fixes

https://github.com/datawire/apro/issues/180

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

File: stable/ambassador/templates/ambassador-pro-service.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      ---
+      apiVersion: ambassador/v1
+      kind: RateLimitService
+      name: ambassador-pro-ratelimit
+      service: 127.0.0.1:{{ .Values.pro.ports.ratelimit }}
+spec:
+  ports:
+  - name: ratelimit-grpc
+    port: 80
+{{- end -}}
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-service.yaml:37]:
```suggestion
{{- end -}}

```

File: stable/ambassador/values.yaml [hunk 2]
```diff
@@ -136,3 +136,12 @@ prometheusExporter:
...
 ambassadorConfig: ""
+
+pro:
+  enabled: false
+  image: quay.io/datawire/ambassador_pro:amb-sidecar-0.4.0
+  ports:
+    auth: 8500
+    ratelimit: 8501
+    ratelimitDebug: 8502
+  licenseKey:
```
Comment (Flydiverny) [inline stable/ambassador/values.yaml]:
```suggestion
  licenseKey:

```

File: stable/ambassador/templates/ambassador-pro-service.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{{- if .Values.pro.enabled -}}
+---
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-service.yaml]:
```suggestion
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  template:
+    metadata:
+      labels:
+        service: ambassador-pro-redis
+    spec:
+      containers:
+      - name: redis
+        image: redis:5.0.1
+      restartPolicy: Always
+{{- end -}}
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-redis.yaml:45]:
```suggestion
{{- end -}}

```

File: stable/ambassador/templates/ambassador-pro-crd.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      served: true
+      storage: true
+  scope: Namespaced
+  names:
+    plural: ratelimits
+    singular: ratelimit
+    kind: RateLimit
+    shortNames:
+    - rl
+{{- end -}}
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-crd.yaml:58]:
```suggestion
{{- end -}}

```

File: stable/ambassador/templates/ambassador-pro-crd.yaml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+{{- if .Values.pro.enabled -}}
+---
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-crd.yaml]:
```suggestion
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{{- if .Values.pro.enabled -}}
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-redis.yaml:1]: If I could would it be preferred to run redis outside of k8s, ie thru aws or other solution?
If yes we should probably allow configuring for this :) Could still default to using provided redis.

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: ambassador-pro-redis
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]: should add missing metadata and use template for names so its possible to install multiple releases, this file and others :) 

```suggestion
  name: {{ include "ambassador.fullname" . }}-pro-redis
  labels:
    app.kubernetes.io/name: {{ include "ambassador.name" . }}
    helm.sh/chart: {{ include "ambassador.chart" . }}
    app.kubernetes.io/instance: {{ .Release.Name }}
    app.kubernetes.io/managed-by: {{ .Release.Service }}
```

ditto on selectors.matchLabels and template labels

File: stable/ambassador/templates/ambassador-pro-crd.yaml [hunk 9]
```diff
@@ -0,0 +1,59 @@
+{{- if .Values.pro.enabled -}}
```
Comment (Flydiverny) [inline stable/ambassador/templates/ambassador-pro-crd.yaml:1]: For the CRDs maybe we should do something similar to [this](https://github.com/helm/charts/blob/master/stable/sealed-secrets/templates/sealedsecret-crd.yaml#L6-L9)
Possibly a flag to create crds or not, but I'm not sure if that's really needed.

**Review by Flydiverny — CHANGES_REQUESTED** [review]

File: stable/ambassador/templates/ambassador-pro-redis.yaml (hunk 7 above)
Comment (iNoahNothing) [reply stable/ambassador/templates/ambassador-pro-redis.yaml:1]: Interesting point. FTR we don't have any working example of doing this in Pro outside of Helm. Might be worth looking into and changing in the future.

File: stable/ambassador/templates/ambassador-pro-redis.yaml (hunk 8 above)
Comment (bpehling) [reply stable/ambassador/templates/ambassador-pro-redis.yaml]: Make sure to get the matchLabels as well
```yaml
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
      app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/ambassador/templates/deployment.yaml [hunk 10]
```diff
@@ -136,6 +136,30 @@ spec:
...
+          - name: REDIS_URL
+            value: ambassador-pro-redis:6379
+          - name: APRO_AUTH_PORT
+            value: "{{ .Values.pro.ports.auth }}"
+          - name: GRPC_PORT
+            value: "{{ .Values.pro.ports.ratelimit }}"
+          - name: DEBUG_PORT
+            value: "{{ .Values.pro.ports.ratelimitDebug }}"
+          - name: AMBASSADOR_LICENSE_KEY 
+            value: {{ .Values.pro.licenseKey }}
```
Comment (bpehling) [inline stable/ambassador/templates/deployment.yaml]: Can this be added as a secret instead?
```suggestion
            valueFrom:
              secretKeyRef:
                name: ambassador-pro-license-key
                key: key
```

Then just add this as a template
``` yaml
---
apiVersion: v1
kind: Secret
metadata:
    name: ambassador-pro-license-key
type: Opaque
data:
    key: {{ .Values.pro.licenseKey | b64enc }}
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml (hunk 8 above)
Comment (bpehling) [reply stable/ambassador/templates/ambassador-pro-redis.yaml]: @Flydiverny 
```yaml
labels:
  app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
```

File: stable/ambassador/templates/deployment.yaml [hunk 11]
```diff
@@ -136,6 +136,30 @@ spec:
...
+            containerPort: {{ .Values.pro.ports.ratelimitDebug }}
+          env:
+          - name: REDIS_SOCKET_TYPE
+            value: tcp
+          - name: REDIS_URL
+            value: ambassador-pro-redis:6379
+          - name: APRO_AUTH_PORT
+            value: "{{ .Values.pro.ports.auth }}"
+          - name: GRPC_PORT
+            value: "{{ .Values.pro.ports.ratelimit }}"
```
Comment (bpehling) [inline stable/ambassador/templates/deployment.yaml:157]: Can log level be configurable as an environment variable?
```yaml
        - name: APP_LOG_LEVEL
          value: "{{ .Values.pro.logLevel }}"
```

File: stable/ambassador/templates/deployment.yaml [hunk 12]
```diff
@@ -136,6 +136,30 @@ spec:
           {{- end }}
           resources:
             {{- toYaml .Values.resources | nindent 12 }}
+        {{- if .Values.pro.enabled }}
+        - name: ambassador-pro
+          image: "{{ .Values.pro.image }}"
```
Comment (bpehling) [inline stable/ambassador/templates/deployment.yaml]:
```suggestion
          image: "{{ .Values.pro.image.repository }}:{{ .Values.pro.image.tag }}"
```

File: stable/ambassador/templates/deployment.yaml (hunk 11 above)
Comment (iNoahNothing) [reply stable/ambassador/templates/deployment.yaml:157]: 👍 Good point

File: stable/ambassador/templates/deployment.yaml (hunk 10 above)
Comment (iNoahNothing) [reply stable/ambassador/templates/deployment.yaml]: yes we can add the option to do both.

File: stable/ambassador/templates/ambassador-pro-crd.yaml (hunk 9 above)
Comment (iNoahNothing) [reply stable/ambassador/templates/ambassador-pro-crd.yaml:1]: The CRDs are required for Ambassador Pro. If you install Ambassador Pro, you must install the CRDs.

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 13]
```diff
@@ -0,0 +1,46 @@
+{{- if .Values.pro.enabled -}}
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ include "ambassador.fullname" . }}-pro-redis
+  labels:
+    app.kubernetes.io/name: {{ include "ambassador.name" . }}
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]:
```suggestion
    app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
    service: ambassador-pro-redis
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+spec:
+  type: ClusterIP
+  ports:
+  - port: 6379
+    targetPort: 6379
+  selector:
+    service: ambassador-pro-redis
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]:
```suggestion
    service: ambassador-pro-redis
    app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
    app.kubernetes.io/instance: {{ .Release.Name }}
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+  labels:
+    app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      service: ambassador-pro-redis
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]:
```suggestion
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+      app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
+      app.kubernetes.io/instance: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        service: ambassador-pro-redis
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]:
```suggestion
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    targetPort: 6379
+  selector:
+    service: ambassador-pro-redis
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{ include "ambassador.fullname" . }}-pro-redis
+  labels:
+    app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml]:
```suggestion
    service: ambassador-pro-redis
    app.kubernetes.io/name: {{ include "ambassador.name" . }}-pro-redis
```

File: stable/ambassador/templates/ambassador-pro-redis.yaml [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+    helm.sh/chart: {{ include "ambassador.chart" . }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/managed-by: {{ .Release.Service }}
```
Comment (bpehling) [inline stable/ambassador/templates/ambassador-pro-redis.yaml:28]:
```suggestion
    app.kubernetes.io/managed-by: {{ .Release.Service }}
    service: ambassador-pro-redis
```

Comment (iNoahNothing) [reply stable/ambassador/templates/ambassador-pro-redis.yaml:28]: What is this change doing?

Comment (bpehling) [reply stable/ambassador/templates/ambassador-pro-redis.yaml:28]: put in the wrong spot, ignore

**Comment by nbkrause** [comment]
@bpehling @Flydiverny The DCO issue with this PR became more effort than it was worth to resolve. I am closing this PR and have created this one with the diff https://github.com/helm/charts/pull/13434

**Pull request closed without merging**